plugin/src/main/java/com/denizenscript/denizen/events/block/MoistureChangeScriptEvent.java

```diff
@@ -0,0 +1,90 @@
+package com.denizenscript.denizen.events.block;
+
+import com.denizenscript.denizen.events.BukkitScriptEvent;
+import com.denizenscript.denizen.objects.LocationTag;
+import com.denizenscript.denizen.objects.MaterialTag;
+import com.denizenscript.denizencore.objects.ObjectTag;
+import com.denizenscript.denizencore.objects.core.ElementTag;
+import org.bukkit.block.data.type.Farmland;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.event.block.MoistureChangeEvent;
+
+public class MoistureChangeScriptEvent extends BukkitScriptEvent implements Listener {
+
+    // <--[event]
+    // @Events
+    // farmland moisture level changes
+    //
+    // @Group Block
+    //
+    // @Location true
+    //
+    // @Triggers when a farmland's moisture level changes.
+    //
+    // @Switch from:<level> to only process the event when the previous moisture level matches the input.
+    // @Switch to:<level> to only process the event when the new moisture level matches the input.
+    //
+    // @Context
+    // <context.location> returns the LocationTag of the farmland.
+    // <context.old_material> returns the MaterialTag the old farmland block, before the level changed.
+    // <context.new_material> returns the MaterialTag the farmland block is changing to.
+    // <context.old_level> returns the ElementTag(Number) of the previous moisture level.
+    // <context.new_level> returns the ElementTag(Number) of the new moisture level.
+    //
+    // @Cancellable true
+    //
+    // @Example
+    // # Announce when farmland begins to dry out.
+    // on farmland moisture level changes from:7 to:6:
+    // - announce "Farmland at location <context.location.simple> lost its water source and began to dry!"
+    //
+    // -->
+
+    public MoistureChangeScriptEvent() {
+        registerCouldMatcher("farmland moisture level changes");
+        registerSwitches("from", "to");
+    }
+
+    public MoistureChangeEvent event;
+    public MaterialTag newMaterial;
+    public LocationTag location;
+    public Farmland oldFarmland;
+    public Farmland newFarmland;
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (!runInCheck(path, location)) {
+            return false;
+        }
+        if (!path.checkSwitch("from", String.valueOf(oldFarmland.getMoisture()))) {
+            return false;
+        }
+        if (!path.checkSwitch("to", String.valueOf(newFarmland.getMoisture()))) {
+            return false;
+        }
+        return super.matches(path);
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        return switch (name) {
+            case "location" -> location;
+            case "old_material" -> new MaterialTag(event.getBlock().getBlockData());
```

Comment:
Isn't this just `oldFarmland`? also there's probably no need to have `newMaterial` in it's own field if it's only used in this context tag (or at least should be consistent with `old_material`)

Reply:
Oh actually yes but no not quite: `oldFarmland` is correct and this code is actively wrong, it's making a read call that can change between ticks from external factors, so eg `after farmland changes: - narrate <context.old_material>` the current code will yield a wrong value